File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.py
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        result = 0
        slow = 0
        n = len(nums)
        fast = 1
        nums.sort()

        while fast < n:
            curr = nums[fast] - nums[slow]

            if curr > k:
                slow += 1
            
            elif curr < k or fast == slow:
                fast += 1
            else:
                result += 1
                slow += 1
                fast += 1

                while fast < n and nums[fast] == nums[fast - 1]:
                    fast += 1
        
        return result
    
#two pointers
# time complexity: O(nlogn)
# space complexity: O(1)

File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.java
class Solution {
    public int findPairs(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        
        HashSet<List<Integer>> result = new HashSet<>();
        
        for(int i = 0; i < nums.length; i++){
            map.put(nums[i], i);
        }

        for(int i = 0; i < nums.length; i++){
            int complement = nums[i] - k;
            if(map.containsKey(complement) && map.get(complement) != i){
                Integer[] temp = {nums[i], complement};
                Arrays.sort(temp);
                result.add(Arrays.asList(temp));
            }

            complement = nums[i] + k;
            if(map.containsKey(complement) && map.get(complement) != i){
                Integer[] temp = {nums[i], complement};
                Arrays.sort(temp);
                result.add(Arrays.asList(temp));
            }
        }

        return result.size();
    }
}